### PR TITLE
started converting client files to es2015 syntax - fix #396

### LINKED ISSUE #396: Use ES2015

### PR DESCRIPTION
**Pull request by elie222**

### CODE REVIEW

File: client/route.js [hunk 1]
```diff
@@ -1,97 +1,106 @@
-Route = function(router, pathDef, options, group) {
-  options = options || {};
+"use strict";
```
Comment (arunoda) [inline client/route.js]: Remove this as I mentioned earlier. We'll use eslint with much better rules.

**Comment by elie222** [comment]
I didn't understand the comment. What's not necessary?

**Comment by elie222** [comment]
But I'm unable to use class without 'use strict'.

**Comment by elie222** [comment]
I get the message:
"Uncaught SyntaxError: Block-scoped declarations (let, const, function, class) not yet supported outside strict mode"
when removing that line.

**Comment by arunoda** [comment]
Okay. You need to add the `ecmascript` package to the project. Try that.

**Comment by elie222** [comment]
Done and works

File: package.js [hunk 2]
```diff
@@ -54,6 +54,7 @@ Package.onTest(function(api) {
 function configure(api) {
   api.versionsFrom('1.0');
```
Comment (arunoda) [inline package.js]: Update this to `1.2`

Comment (elie222) [inline package.js]: Done.

**Comment by elie222** [comment]
Well I learnt a few things today. I should've branched off of the `ssr` branch. Need to make sure the tests are passing when I make a commit. And I probably shouldn't edit code without testing on a working Meteor app.

But the project should now be fully converted to ES2015.

**Comment by arunoda** [comment]
This is great. If you like you can create a different branch.

File: server/plugins/ssr_data.js [hunk 3]
```diff
@@ -1,47 +1,53 @@
...
   }
 
-  if(originalSubscribe) {
+  if (originalSubscribe) {
     originalSubscribe.apply(this, arguments);
   }
-  return {ready: function () {return true}};
+  
+  return {
+    ready: () => { return true; }
```
Comment (arunoda) [inline server/plugins/ssr_data.js]: Do like this: `ready: () => true`

File: server/plugins/fast_render.js [hunk 4]
```diff
@@ -1,36 +1,34 @@
...
-        var queryParams = routeParams.query;
-        var params = _.omit(routeParams, 'query');
+      if (route.subscriptions) {
+        const queryParams = routeParams.query;
+        const params = _.omit(routeParams, 'query');
         route.subscriptions(params, queryParams);
       }
-      _.each(route._subsMap, function (args) {
-        self.subscribe.apply(self, args);
+      route._subsMap.forEach((args) => {
```
Comment (arunoda) [inline server/plugins/fast_render.js]: If there is a single argument you don't need parenthesis. You can like like 

```
args => {
  ...
}
```

File: server/plugins/ssr_data.js [hunk 5]
```diff
@@ -52,13 +58,12 @@ Tracker.autorun = function (fn) {
...
-  var original = Meteor[methodName];
-  Meteor[methodName] = function() {
-    var self = this;
-    var args = arguments;
-    var response = FlowRouter.ssrContext.withValue(null, function() {
-      return original.apply(self, args);
+['call', 'apply'].forEach((methodName) => {
+  const original = Meteor[methodName];
+  Meteor[methodName] = () => {
+    const args = arguments;
```
Comment (arunoda) [inline server/plugins/ssr_data.js]: This args will be incorrect. Get it like this:

```js
Meteor[methodName] = (...args) => {
  
}
```

File: server/plugins/fast_render.js [hunk 6]
```diff
@@ -1,36 +1,34 @@
...
-    FastRender.route(route.pathDef, function (routeParams, path) {
-      var self = this;
-
+  FlowRouter._routes.forEach((route) => {
+    FastRender.route(route.pathDef, (routeParams, path) => {
       // anyone using Meteor.subscribe for something else?
-      var original = Meteor.subscribe;
-      Meteor.subscribe = function () {
+      const original = Meteor.subscribe;
+      Meteor.subscribe = () => {
```
Comment (arunoda) [inline server/plugins/fast_render.js]: Get args using `...args`

File: server/plugins/ssr_data.js [hunk 7]
```diff
@@ -1,47 +1,53 @@
-var originalSubscribe = Meteor.subscribe;
-Meteor.subscribe = function(pubName) {
-  var params = Array.prototype.slice.call(arguments, 1);
+const originalSubscribe = Meteor.subscribe;
 
-  var ssrContext = FlowRouter.ssrContext.get();
-  if(ssrContext) {
-    FlowRouter.inSubscription.withValue(true, function() {
+Meteor.subscribe = function (pubName) {
```
Comment (arunoda) [inline server/plugins/ssr_data.js]: You can get params like this too: `pubName, ...params`

But, we need to use `arguments` below. So, I'm okay with this too.

File: server/route.js [hunk 8]
```diff
@@ -1,195 +1,196 @@
...
+    this._cache[url] = info;
   }
-
-  var info = {
-    data: data,
-    timeoutHandle: setTimeout(function() {
-      delete self._cache[url];
-    }, timeout)
+  
+  register(name, sub, options) {
```
Comment (arunoda) [inline server/route.js]: I think we can remove following three methods.

File: server/router.js [hunk 9]
```diff
@@ -1,133 +1,162 @@
...
+  }
+  
+  getRouteName() {
+    const current = this.current();
+    if(current) {
+      return current.route.name;
+    }
+  }
+  
+  ready() {
```
Comment (arunoda) [inline server/router.js]: Remove following three methods.

**Comment by arunoda** [comment]
For SSR, we don't have tests yet. Try to play with this app: https://github.com/arunoda/hello-react-meteor to check whether it's working or not.

File: server/route.js (hunk 8 above)
Comment (elie222) [inline server/route.js]: You mean these 3?
```
  register(name, sub, options) {
    this._subsMap[name] = sub;
  };
  
  subscription(name) {
    return this._subsMap[name];
  }
  
  middleware(middleware) {
  }
```

In the demo app I was using it caused an error. This app: https://github.com/timbrandin/meteor-react-flowrouter-jsx-templating-example

Comment (arunoda) [inline server/route.js]: Yep.
Use this app. See: https://github.com/arunoda/hello-react-meteor
In this version of SSR, we don't have subscription registration support.

**Comment by elie222** [comment]
Played around with it in the demo app and it's working. Made the changes recommended above. Apart from removing the 3 methods that caused the app to crash server side

**Comment by arunoda** [comment]
Okay. I'll have a check on what's going on here.

**Comment by arunoda** [comment]
Okay. Had some bugs in the `server/route.js`. Apply this patch code

```patch
diff --git a/server/route.js b/server/route.js
index 8d1c79e..eba76b7 100644
--- a/server/route.js
+++ b/server/route.js
@@ -7,54 +7,50 @@ Route = class {
     this.options = options;
     this.name = options.name;
     this.pathDef = pathDef;
-  
-  
+
     // Route.path is deprecated and will be removed in 3.0
     this.path = this.pathDef;
-  
+
     this.action = options.action || Function.prototype;
     this._router = router;
-    this.subscriptions = options.subscriptions || Function.prototype;
-    this._subsMap = {};
     this._cache = {};
-  
+
     Picker.middleware(Npm.require('connect').cookieParser());
     // process null subscriptions with FR support
     Picker.middleware(FastRender.handleOnAllRoutes);
-    
+
     const route = FlowRouter.basePath + this.pathDef;
     Picker.route(route, (params, req, res, next) => {
-  
       if (!this.isHtmlPage(req.url)) {
         return next();
       }
-  
+
       const cachedPage = this._lookupCachedPage(req.url);
       if (cachedPage) {
         return this._processFromCache(cachedPage, res, next);
       }
-  
+
       const processFromSsr = this._processFromSsr.bind(this, params, req, res);
       FastRender.handleRoute(processFromSsr, params, req, res, (data) => {
         next();
-      }); 
+      });
     });
   }
-  
+
   _processFromSsr(params, req, res) {
     const self = this;
     const ssrContext = new SsrContext();
-    
-    this._router.ssrContext.withValue(ssrContext, () => {
+
+    self._router.ssrContext.withValue(ssrContext, function() {
       const queryParams = params.query;
       // We need to remove `.query` since it's not part of our params API
-      // But we only need to remove it in our copy. 
+      // But we only need to remove it in our copy.
       // We should not trigger any side effects
       params = _.clone(params);
       delete params.query;
-      const context = this._buildContext(req.url, params, queryParams);
-  
-      this._router.currentRoute.withValue(context, () => {
+      const context = self._buildContext(req.url, params, queryParams);
+
+      self._router.currentRoute.withValue(context, function() {
         try {
           // get the data for null subscriptions and add them to the
           // ssrContext
@@ -62,20 +58,16 @@ Route = class {
           if(frData) {
             ssrContext.addData(frData.collectionData);
           }
-  
-          if(this.options.subscriptions) {
-            this.options.subscriptions.call(this, params, queryParams);
-          }
-  
-          if(this.options.action) {
-            this.options.action.call(this, params, queryParams);
+
+          if(self.options.action) {
+            self.options.action.call(self, params, queryParams);
           }
         } catch(ex) {
           console.error("Error when doing SSR. path:", req.url, " ", ex.message);
           console.error(ex.stack);
         }
       });
-  
+
       const originalWrite = res.write;
       res.write = function (data) {
         if(typeof data === 'string') {
@@ -83,52 +75,52 @@ Route = class {
           if(head && head.trim() !== "") {
             data = data.replace('</head>', head + '\n</head>');
           }
-  
+
           const reactRoot = ssrContext.getHtml();
           if (self._router.deferScriptLoading) {
             data = moveScripts(data);
           }
           data = data.replace('<body>', '<body>' + reactRoot);
-  
+
           const pageInfo = {
             frData: res.getData("fast-render-data"),
             html: data
           };
-  
+
           // cache the page if mentioned a timeout
           if(self._router.pageCacheTimeout) {
             self._cachePage(req.url, pageInfo, self._router.pageCacheTimeout);
           }
         }
-        
+
         originalWrite.call(this, data);
       };
     });
-  
+
     function moveScripts(data) {
       const $ = Cheerio.load(data, {
         decodeEntities: false
       });
       const heads = $('head script');
       $('body').append(heads);
-  
+
       // Remove empty lines caused by removing scripts
       $('head').html($('head').html().replace(/(^[ \t]*\n)/gm, ''));
-  
+
       return $.html();
     }
   }
-  
+
   _processFromCache(pageInfo, res, next) {
     const originalWrite = res.write;
-    res.write = (data) => {
+    res.write = function(data) {
       originalWrite.call(this, pageInfo.html);
     }
-  
+
     res.pushData('fast-render-data', pageInfo.frData);
     next();
   }
-  
+
   _buildContext(url, params, queryParams) {
     const context = {
       route: this,
@@ -136,61 +128,49 @@ Route = class {
       params: params,
       queryParams: queryParams
     };
-  
+
     return context;
   }
-  
+
   isHtmlPage(url) {
     const pathname = Url.parse(url).pathname;
     const ext = pathname.split('.').slice(1).join('.');
-  
+
     // if there is no extention, yes that's a html page
     if(!ext) {
       return true;
     }
-  
+
     // if this is htm or html, yes that's a html page
     if(/^htm/.test(ext)) {
       return true;
     }
-  
+
     // if not we assume this is not as a html page
     // this doesn't do any harm. But no SSR
     return false;
   }
-  
+
   _lookupCachedPage(url) {
     const info = this._cache[url];
     if(info) {
       return info.data;
     }
   }
-  
+
   _cachePage(url, data, timeout) {
     const existingInfo = this._cache[url];
     if(existingInfo) {
       clearTimeout(existingInfo.timeoutHandle);
     }
-  
+
     const info = {
       data: data,
       timeoutHandle: setTimeout(() => {
         delete this._cache[url];
       }, timeout)
     };
-  
+
     this._cache[url] = info;
   }
-  
-  register(name, sub, options) {
-    this._subsMap[name] = sub;
-  };
-  
-  subscription(name) {
-    return this._subsMap[name];
-  }
-  
-  middleware(middleware) {
-  
-  }
 }

```

**Comment by elie222** [comment]
Working through the code trying to understand it better and wondering whether these lines in server/router.js are necessary?
```
    this.subscriptions = Function.prototype;
    this.inSubscription = new Meteor.EnvironmentVariable();
```

**Comment by arunoda** [comment]
We use `inSubscription` because of this.

See here: https://github.com/kadirahq/flow-router/blob/ssr/server/plugins/ssr_data.js#L24
We are starting a context for SSR data. So, the UI can only get the data it only subscribe to even in the server when rendering the UI.
But when the UI invoke the subscription on the server, publication needs to access the whole DB. We need to allow that, this is how we do it.

**Comment by arunoda** [comment]
This code looks pretty great to me. Do you have anything left to do?

Then send me a another PR with [squashing](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/) the changes.

**Comment by elie222** [comment]
i made another pull request from another branch

**Comment by arunoda** [comment]
Fixed by: #411

**Pull request closed without merging**